src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -417,6 +417,84 @@ class AutoDiffBroadcastInDimRev
                           MGradientUtilsReverse *gutils) const {}
 };
 
+class AutoDiffDynamicSliceUpdateRev
+    : public ReverseAutoDiffOpInterface::ExternalModel<
+          AutoDiffDynamicSliceUpdateRev, DynamicUpdateSliceOp> {
+public:
+  LogicalResult createReverseModeAdjoint(Operation *orig, OpBuilder &builder,
+                                         MGradientUtilsReverse *gutils,
+                                         SmallVector<Value> caches) const {
+    auto op = cast<DynamicUpdateSliceOp>(orig);
+    auto operand = op.getOperand();
+    auto update = op.getUpdate();
+
+    SmallVector<Value> startIndices;
+
+    Value diffe;
+    if (!gutils->isConstantValue(operand)) {
+
+      for (auto cache : caches) {
+        startIndices.push_back(gutils->popCache(cache, builder));
+      }
+
+      diffe = gutils->diffe(orig->getResult(0), builder);
+      auto operandDiffe = builder.create<DynamicUpdateSliceOp>(
+          orig->getLoc(), diffe,
+          cast<AutoDiffTypeInterface>(update.getType())
+              .createNullValue(builder, orig->getLoc()),
+          startIndices);
+
+      gutils->addToDiffe(operand, operandDiffe, builder);
+    }
+
+    if (!gutils->isConstantValue(update)) {
+      if (startIndices.size() != caches.size()) {
+        for (auto cache : caches) {
+          startIndices.push_back(gutils->popCache(cache, builder));
+        }
+      }
+
+      if (!diffe)
+        diffe = gutils->diffe(orig->getResult(0), builder);
+
+      auto sliceSizes = builder.getDenseI64ArrayAttr(
+          cast<TensorType>(update.getType()).getShape());
+      auto updateDiffe = builder.create<DynamicSliceOp>(
+          orig->getLoc(), diffe, startIndices, sliceSizes);
+
+      gutils->addToDiffe(update, updateDiffe, builder);
+    }
+
+    if (diffe)
+      gutils->zeroDiffe(op, builder);
+
+    return success();
+  }
+
+  SmallVector<Value> cacheValues(Operation *orig,
+                                 MGradientUtilsReverse *gutils) const {
+    auto op = cast<DynamicUpdateSliceOp>(orig);
+    Operation *newOp = gutils->getNewFromOriginal(orig);
+    OpBuilder cacheBuilder(newOp);
+
+    auto startIndices = op.getStartIndices();
+
+    SmallVector<Value> caches;
+    caches.reserve(startIndices.size());
+
```

Comment:
this should have a check where we only push the caches if either operand is active

Reply:
done. but hopefully the interface is not called in this case due to the op being constant:

https://github.com/EnzymeAD/Enzyme/blob/f8a48926b72284c8d8c6e892fcd2c44191ef06a9/enzyme/Enzyme/MLIR/Interfaces/EnzymeLogicReverse.cpp#L43-L47